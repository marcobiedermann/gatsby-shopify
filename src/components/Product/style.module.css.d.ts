declare const styles: {
  readonly product: string;
  readonly product__description: string;
  readonly product__image: string;
  readonly product__images: string;
  readonly product__options: string;
  readonly product__title: string;
};
export = styles;
