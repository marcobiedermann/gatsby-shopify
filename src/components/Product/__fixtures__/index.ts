import { ProductProps } from '..';

const product: ProductProps = {
  availableForSale: true,
  description: 'Product Description',
  handle: 'product-handle',
  images: [
    {
      id: '12345678-9012-3456-7890-123456789012',
      originalSrc: '/path/to/image.jpg',
    },
  ],
  options: [
    {
      name: 'Size',
      shopifyId: '12345678-9012-3456-7890-123456789012',
      values: ['s', 'm', 'l'],
    },
  ],
  priceRange: {
    minVariantPrice: {
      amount: '100',
      currencyCode: 'EUR',
    },
    maxVariantPrice: {
      amount: '100',
      currencyCode: 'EUR',
    },
  },
  shopifyId: '12345678-9012-3456-7890-123456789012',
  tags: ['tag'],
  title: 'Product Title',
  vendor: 'Vendor',
};

export default product;
