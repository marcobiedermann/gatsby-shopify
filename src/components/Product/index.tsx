import { Field, Form, Formik } from 'formik';
import React, { FC } from 'react';
import Button from '../Button';
import Image from '../Image';
import Label from '../Label';
import Tags from '../Tags';
import styles from './style.module.css';

interface Image {
  id: string;
  originalSrc: string;
}

interface Option {
  name: string;
  shopifyId: string;
  values: string[];
}

export interface ProductProps {
  availableForSale: boolean;
  description: string;
  handle: string;
  images: Image[];
  options: Option[];
  priceRange: {
    minVariantPrice: {
      amount: string;
      currencyCode: string;
    };
    maxVariantPrice: {
      amount: string;
      currencyCode: string;
    };
  };
  shopifyId: string;
  tags: string[];
  title: string;
  vendor: string;
}

const Product: FC<ProductProps> = (props) => {
  const { description, images, options, priceRange, tags, title, vendor } = props;

  return (
    <div className={styles.product}>
      <div style={{ width: '50%' }}>
        <ul className={styles.product__images}>
          {images.map((image) => (
            <li key={image.id}>
              <Image className={styles.product__image} src={image.originalSrc} alt={title} />
            </li>
          ))}
        </ul>
      </div>
      <div style={{ paddingLeft: '6em', paddingRight: '6em', width: '50%' }}>
        <h1 className={styles.product__title}>{title}</h1>
        <h2>by {vendor}</h2>
        <div>
          {priceRange.minVariantPrice.amount} {priceRange.minVariantPrice.currencyCode}
        </div>
        <div className={styles.product__description}>
          <p>{description}</p>
        </div>
        <Tags tags={tags} />
        <Formik
          initialValues={{
            quantity: 1,
          }}
          onSubmit={(values, actions) => {
            console.log({ values });
            actions.setSubmitting(false);
          }}
        >
          {({ isSubmitting }) => (
            <Form>
              {options.map((option) => (
                <div className="form__field" key={option.shopifyId}>
                  {option.values.length > 5 ? (
                    <>
                      <Label htmlFor={option.name}>{option.name}</Label>
                      <Field
                        as="select"
                        className="form__input"
                        id={option.name}
                        name={option.name}
                      >
                        {option.values.map((value) => (
                          <option key={`${option.name}-${value}`} value={value}>
                            {value}
                          </option>
                        ))}
                      </Field>
                    </>
                  ) : (
                    <fieldset>
                      <legend>{option.name}</legend>
                      <div className={styles.product__options}>
                        {option.values.map((value) => (
                          <div key={value}>
                            <Label htmlFor={value}>{value}</Label>
                            <Field type="radio" id={value} name={option.name} value={value} />
                          </div>
                        ))}
                      </div>
                    </fieldset>
                  )}
                </div>
              ))}
              <div className="form__field">
                <Label htmlFor="quantity">Quantity</Label>
                <Field
                  className="form__input"
                  id="quantity"
                  min="1"
                  name="quantity"
                  step="1"
                  type="number"
                />
              </div>
              <div className="form__field">
                <Button type="submit" disabled={isSubmitting}>
                  Add to cart
                </Button>
              </div>
            </Form>
          )}
        </Formik>
      </div>
    </div>
  );
};

export default Product;
