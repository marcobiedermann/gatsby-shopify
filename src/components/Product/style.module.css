.product {
  display: flex;
}

.product__description {
  margin-bottom: 1.5em;
}

.product__image {
  width: 480px;
}

.product__images {
  display: grid;
  gap: 1em;
  grid-template-columns: repeat(2, 1fr);
  list-style: none;
  margin: 0;
  padding: 0;
}

.product__options {
  display: flex;
  gap: 1em;
}

.product__title {
  font-size: 2rem;
}
