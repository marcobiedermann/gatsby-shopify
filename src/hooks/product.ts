import { graphql, useStaticQuery } from 'gatsby';

interface Image {
  originalSrc: string;
}

interface PriceRange {
  minVariantPrice: VariantPrice;
}

interface ShopifyProductQuery {
  handle: string;
  images: Image[];
  priceRange: PriceRange;
  shopifyId: string;
  title: string;
}

interface VariantPrice {
  amount: string;
  currencyCode: string;
}

interface AllShopifyProductQuery {
  allShopifyProduct: {
    edges: {
      node: ShopifyProductQuery;
    }[];
  };
}

export function useAllShopifyProduct() {
  return useStaticQuery<AllShopifyProductQuery>(graphql`
    {
      allShopifyProduct(sort: { fields: [title] }) {
        edges {
          node {
            handle
            images {
              originalSrc
            }
            priceRange {
              minVariantPrice {
                amount
                currencyCode
              }
            }
            shopifyId
            title
          }
        }
      }
    }
  `);
}
