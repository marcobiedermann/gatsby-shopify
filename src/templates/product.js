import React from "react"
import Layout from "../components/layout"

const ProductTemplate = ({ pageContext }) => {
  const { product } = pageContext

  return (
    <Layout>
      <h1>{product.title}</h1>
      <div>{product.description}</div>
    </Layout>
  )
}

export default ProductTemplate
