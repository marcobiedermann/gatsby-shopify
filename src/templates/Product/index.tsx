import { graphql, PageProps } from 'gatsby';
import React, { FC } from 'react';
import Layout from '../../components/Layout';
import Product from '../../components/Product';

interface Image {
  id: string;
  originalSrc: string;
}

interface Option {
  name: string;
  shopifyId: string;
  values: string[];
}

export interface DataProps {
  shopifyProduct: {
    availableForSale: boolean;
    description: string;
    handle: string;
    images: Image[];
    options: Option[];
    priceRange: {
      minVariantPrice: {
        amount: string;
      };
      maxVariantPrice: {
        amount: string;
      };
    };
    shopifyId: string;
    tags: string[];
    title: string;
  };
}

const ProductTemplate: FC<PageProps<DataProps>> = (props) => {
  const {
    data: { shopifyProduct },
    location,
  } = props;

  return (
    <Layout location={location}>
      <Product {...shopifyProduct} />
    </Layout>
  );
};

export const pageQuery = graphql`
  query ProductByShopifyId($shopifyId: String!) {
    shopifyProduct(shopifyId: { eq: $shopifyId }) {
      availableForSale
      description
      handle
      images {
        id
        originalSrc
      }
      options {
        shopifyId
        name
        values
      }
      priceRange {
        minVariantPrice {
          amount
        }
        maxVariantPrice {
          amount
        }
      }
      shopifyId
      tags
      title
    }
  }
`;

export default ProductTemplate;
