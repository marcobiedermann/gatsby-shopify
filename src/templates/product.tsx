import React, { FC } from "react"
import Layout from "../components/Layout"
import Product, { ProductProps } from "../components/Product"

export interface ProductTemplateProps {
  pageContext: {
    product: ProductProps
  }
}

const ProductTemplate: FC<ProductTemplateProps> = props => {
  const { pageContext } = props
  const { product } = pageContext

  return (
    <Layout>
      <Product {...product} />
    </Layout>
  )
}

export default ProductTemplate
