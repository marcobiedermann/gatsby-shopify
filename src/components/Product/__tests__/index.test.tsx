import { render } from '@testing-library/react';
import React from 'react';
import Product from '..';
import product from '../__fixtures__';

describe('product component', () => {
  it('should render correctly', () => {
    const { asFragment } = render(<Product {...product} />);

    expect(asFragment()).toMatchSnapshot();
  });
});
