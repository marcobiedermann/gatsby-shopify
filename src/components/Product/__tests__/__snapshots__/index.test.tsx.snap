// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`product component should render correctly 1`] = `
<DocumentFragment>
  <div
    class="product"
  >
    <div
      style="width: 50%;"
    >
      <ul
        class="product__images"
      >
        <li>
          <img
            alt="Product Title"
            class="product__image"
            src="/path/to/image.jpg"
          />
        </li>
      </ul>
    </div>
    <div
      style="padding-left: 6em; padding-right: 6em; width: 50%;"
    >
      <h1
        class="product__title"
      >
        Product Title
      </h1>
      <h2>
        by Vendor
      </h2>
      <div>
        100 EUR
      </div>
      <div
        class="product__description"
      >
        <p>
          Product Description
        </p>
      </div>
      <ul
        class="tags"
      >
        <li>
          <span>
            tag
          </span>
        </li>
      </ul>
      <form
        action="#"
      >
        <div
          class="form__field"
        >
          <fieldset>
            <legend>
              Size
            </legend>
            <div
              class="product__options"
            >
              <div>
                <label
                  class="label"
                  for="s"
                >
                  s
                </label>
                <input
                  id="s"
                  name="Size"
                  type="radio"
                  value="s"
                />
              </div>
              <div>
                <label
                  class="label"
                  for="m"
                >
                  m
                </label>
                <input
                  id="m"
                  name="Size"
                  type="radio"
                  value="m"
                />
              </div>
              <div>
                <label
                  class="label"
                  for="l"
                >
                  l
                </label>
                <input
                  id="l"
                  name="Size"
                  type="radio"
                  value="l"
                />
              </div>
            </div>
          </fieldset>
        </div>
        <div
          class="form__field"
        >
          <label
            class="label"
            for="quantity"
          >
            Quantity
          </label>
          <input
            class="form__input"
            id="quantity"
            min="1"
            name="quantity"
            step="1"
            type="number"
            value="1"
          />
        </div>
        <div
          class="form__field"
        >
          <button
            class="button"
            type="submit"
          >
            Add to cart
          </button>
        </div>
      </form>
    </div>
  </div>
</DocumentFragment>
`;
